pkg/controller/deployment/deployment_controller.go

```diff
@@ -438,41 +438,124 @@ func (dc *DeploymentController) syncDeployment(key string) error {
+// of the new replica set and scaling down can decrease the sizes of the old ones, both of which would
+// have the effect of hastening the rollout progress, which could produce a higher proportion of unavailable
+// replicas in the event of a problem with the rolled out template. Should run only on scaling events or
+// when a deployment is paused and not during the normal rollout process.
+func (dc *DeploymentController) scale(deployment *extensions.Deployment, newRS *extensions.ReplicaSet, oldRSs []*extensions.ReplicaSet) error {
+	// Paused and recreate deployments may not have a new replica set. Additionally,
+	// in case no old replica sets have pods, the controller should scale the new
+	// replica set to the full count.
+	if newRS != nil && len(controller.FilterActiveReplicaSets(oldRSs)) == 0 {
+		if newRS.Spec.Replicas == deployment.Spec.Replicas {
+			return nil
+		}
+		setReplicaAnnotations(newRS, deployment.Spec.Replicas, deployment.Spec.Replicas)
+		if _, _, err := dc.scaleReplicaSetAndRecordEvent(newRS, deployment.Spec.Replicas, deployment); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	// If the new replica set is saturated, old replica sets should be fully scaled down.
+	// This case handles replica set adoption during a saturated new replica set.
+	if deploymentutil.IsSaturated(deployment, newRS) {
+		for _, old := range controller.FilterActiveReplicaSets(oldRSs) {
+			setReplicaAnnotations(old, deployment.Spec.Replicas, deployment.Spec.Replicas)
+			if _, _, err := dc.scaleReplicaSetAndRecordEvent(old, 0, deployment); err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+
+	// There are old replica sets with pods and the new replica set is not saturated.
+	// We need to proportionally scale all replica sets (new and old) in case of a
+	// rolling deployment.
+	if deploymentutil.IsRollingUpdate(deployment) {
+		allRSs := append(oldRSs, newRS)
+		allRSs = controller.FilterActiveReplicaSets(allRSs)
+
+		totalReplicas := deploymentutil.GetReplicaCountForReplicaSets(allRSs)
+		totalAdded := int32(0)
+		// Number of additional pods that can be either added or removed from the total pod count.
+		additional := deployment.Spec.Replicas - totalReplicas
+		if additional == 0 {
+			return nil
+		}
+
+		if additional < 0 {
+			// When scaling down replica sets with the same size, scale down older first.
+			sort.Sort(controller.ReplicaSetsBySizeOlder(allRSs))
+		} else {
+			// When scaling up replica sets with the same size, scale up newer first.
+			sort.Sort(controller.ReplicaSetsBySizeNewer(allRSs))
+		}
+
+		for i := range allRSs {
+			rs := allRSs[i]
+			setReplicaAnnotations(rs, deployment.Spec.Replicas, deployment.Spec.Replicas+maxSurge(*deployment))
+
+			proportion := estimateProportion(rs, additional, totalReplicas, totalAdded)
+			if proportion == 0 {
+				// Keep replica set annotations in sync with the latest state of the deployment.
+				dc.updateRS(rs)
```

Comment:
I agree it is better if all the RS annotations are updated.  By default do we never GC replica sets?  Would that mean every scale event needs to update a RS for every rollout ever made?  Would be nice to avoid that if possible.  We should also make sure we perform only 1 write to each RS instead of updating the replicaCount and annotation separately.